### DISCUSSION TITLE
eglot-format-and-organize-imports: format a file and organize its imports

### DISCUSSION THREAD

**Discussion by adonovan** [joaotavora/eglot#1409] [General]
This function augments `eglot-format-buffer` so that it in addition to formatting, it organizes the imports too:

```elisp
(defun eglot-format-and-organize-imports ()
  "Update the imports and format the file, analogous to commands like Go's `goimports`
but agnostic to language, mode, and server."
  (interactive)
  (let* ((actions (eglot-code-actions (point-min) (point-max) "source.organizeImports")))
    (when actions
      (eglot-execute (eglot--current-server-or-lose) (car actions)))
    (eglot-format-buffer)))
```

You can add it to your favorite mode hook like so:

```elisp
(add-hook 'go-mode-hook
 (lambda ()
   (add-hook 'before-save-hook #'eglot-format-and-organize-imports nil t)))
```

In Go, it's the norm for most users to apply these two operations on save, and I assume that must be true in many other languages too. Does this seems like a useful piece of core functionality for the eglot package?

**Comment by manuel-uberti** [comment]
It may be worth noticing that Eglot already provides `eglot-code-action-organize-imports`, so a command addressing that and `eglot-format-buffer` is easier to add to a user's own configuration.

**Reply by stapelberg** [reply]
FWIW, here’s what the gopls docs (at https://github.com/golang/tools/blob/master/gopls/doc/emacs.md#organizing-imports-with-eglot) currently recommend:

```lisp
(add-hook 'before-save-hook
    (lambda ()
        (call-interactively 'eglot-code-action-organize-imports))
    nil t)
```

This is *almost* convenient. 

`eglot-format-buffer` is a little easier to configure as a hook compared to `eglot-code-action-organize-imports` because the latter needs to be called interactively.

Maybe we just need to provide a function that amounts to `eglot-code-action-organize-imports`, but does not need to be called interactively and can therefore easily be added as a hook?

**Reply by joaotavora** [reply]
> Maybe we just need to provide a function that amounts to eglot-code-action-organize-imports, but does not need to be called interactively and can therefore easily be added as a hook?

Every function can be called non-interactively :-) and this one is no exception:
```
(add-hook 'before-save-hook
    (lambda ()
        (eglot-code-action-organize-imports (point-min) (point-max))
    nil t)
```

Is probably what you're after.

You could argue that `(add-hook 'before-save-hook 'eglot-code-action-organize-imports)` would be cleaner, and I agree, but it doesn't work at the moment because the first `BEG` argument to these "code actions" functions is required, and even if it wasn't we;d have to default BEG and END  to `(point-min)` and `(point-max)` non-interactively, which could be a backward incompatible change.

It could be done for some actions though.  But I'm getting ahead of myself: likely the above snippet is more than enough.

**Reply by stapelberg** [reply]
Thanks for the details. 

Sure, the snippet works. But it strikes me as rather complicated for something that virtually every Go programmer should be using. 

As `use-package` is now included in Emacs as well (and is generally recommended pretty widely), I would like to use its `:hook` directive (see https://www.gnu.org/software/emacs/manual/html_node/use-package/Hooks.html) to configure go-mode and all the hooks I want to have active in go-mode. 

For reference, this is an example setup right now:

```lisp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Enable gopls via Eglot (included in Emacs).
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; The depth of -10 places this before eglot's willSave notification,
;; so that that notification reports the actual contents that will be saved.
(defun eglot-format-buffer-on-save ()
  (add-hook 'before-save-hook #'eglot-format-buffer -10 t))

(defun eglot-organize-imports-on-save ()
  (call-interactively 'eglot-code-action-organize-imports))

(defun eglot-organize-imports-on-save-hook ()
  (add-hook 'before-save-hook #'eglot-organize-imports-on-save -10 t))

(use-package eglot
  :hook
  (go-ts-mode . eglot-ensure)
  ;; Optional: format Go code when saving buffers.
  (go-ts-mode . eglot-format-buffer-on-save)
  ;; Optional: organize imports (goimports) when saving buffers.
  (go-ts-mode . eglot-organize-imports-on-save-hook))
```

It would be great if we could provide a function that users can just call with `use-package`’s `:hook` directive. Speaking purely from the user perspective, something like this is what I’m looking for:

```lisp
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Enable gopls via Eglot (included in Emacs).
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(use-package eglot
  :hook
  (go-ts-mode . eglot-ensure)
  ;; Optional: format Go code when saving buffers.
  (go-ts-mode . eglot-hook-format-buffer-on-save)
  ;; Optional: organize imports (goimports) when saving buffers.
  (go-ts-mode . eglot-hook-organize-imports-on-save))
```

**Reply by joaotavora** [reply]
Sorry, I don't this buy this declarative obfuscating-macro fashion, never did, so there's no point in trying to sell it to me.  believe in doing things by composition of modular features, not cartesian multiplication of every feature between itself.  What if you want to organize things, on "mouse leaves buffer", or any of the million hooks around? "On save" isn't even defined: before and after save is.

Also  I'll take a bit of Lisp with well defined rules than some macro DSL which forces me to learn about rules (avoid submitting bug reports with `use-package`, btw. And please stop promoting `eglot-ensure`, see the manual why.).  

Anyway, the argless `eglot-code-action-organize-imports` which defaults to the whole buffer is as far as I will go.  You can then put it in `before-save-hook`, `after-save-hook`, `my-sunday-hook`, etc.  With `add-hook` preferably, but what you do in your config is not my business (unless you ask me to debug it, then I will ask you to remove the cruft).

**Reply by stapelberg** [reply]
I sent https://go.dev/cl/592335 to update our docs regarding the “before-save” terminology.

I don’t understand what you are referring to with regards to `eglot-ensure`. The Emacs manual clearly documents `eglot-ensure` without any warnings or discouragement: https://www.gnu.org/software/emacs/manual/html_node/eglot/Starting-Eglot.html

In general, I am disappointed to learn that `use-package` apparently is contentious among Emacs maintainers. When I first discovered `use-package`, it has helped me make Emacs start up faster and structure my configuration better. I don’t think telling eglot users to avoid `use-package` is a good thing :(

**Reply by manuel-uberti** [reply]
> I don’t understand what you are referring to with regards to `eglot-ensure`. The Emacs manual clearly documents `eglot-ensure` without any warnings or discouragement: https://www.gnu.org/software/emacs/manual/html_node/eglot/Starting-Eglot.html

There is the following in that manual page, which should be taken into account when debugging problems:

_Note that this variant of starting an Eglot session is non-interactive, so it should be used only when you are confident that Eglot can be started reliably for any file which may be visited with the major-mode in question._
 
> In general, I am disappointed to learn that `use-package` apparently is contentious among Emacs maintainers. When I first discovered `use-package`, it has helped me make Emacs start up faster and structure my configuration better. I don’t think telling eglot users to avoid `use-package` is a good thing :(

I think the issue is simply that not everyone uses `use-package` (I don't, for instance), so taking it for granted when reporting bugs may be a misleading assumption.

**Reply by stapelberg** [reply]
> Note that this variant of starting an Eglot session is non-interactive, so it should be used only when you are confident that Eglot can be started reliably for any file which may be visited with the major-mode in question.

I don’t know what this sentence is trying to convey (it sounds like there might be a problem with error handling?) or what I should be doing instead. 

> I think the issue is simply that not everyone uses `use-package` (I don't, for instance), so taking it for granted when reporting bugs may be a misleading assumption.

The way I understood João’s message was that I should not file bug reports if I’m using `use-package`. I’m not taking anything for granted, nor am I saying everyone should use `use-package`. I just think using a popular Emacs builtin package should not be a reason to not be allowed to report bugs.

**Reply by manuel-uberti** [reply]
> I don’t know what this sentence is trying to convey (it sounds like there might be a problem with error handling?) or what I should be doing instead.

You should first try with just `M-x eglot` in the buffer visiting the file you are interested in.
 
> The way I understood João’s message was that I should not file bug reports if I’m using `use-package`. I’m not taking anything for granted, nor am I saying everyone should use `use-package`. I just think using a popular Emacs builtin package should not be a reason to not be allowed to report bugs.

Well, if your bug report contains `use-package` statements *and* the maintainer(s) (or just a regular user as myself) are not used to this macro, you are asking the maintainer(s) to reproduce a problem with something new to them which they have to learn. It's an extra burden, in my opinion. `use-package` is popular but that does not mean that everyone should learn it in order to fix bugs.

**Reply by joaotavora** [reply]
> I don’t know what this sentence is trying to convey (it sounds like there might be a problem with error handling?) or what I should be doing instead.

Indeed to make it _crystal clear_ it should say "started reliably  with M-x eglot"

> you are asking the maintainer(s) to reproduce a problem with something new to them which they have to learn.

Yes, and it's a pretty big language, judging by the docstring and manual. Even if it isn't new,  it needlessly obfuscates and makes reproduction harder.  You're asking the maintainer's brain to macroexpand those forms. It's like adding details about bathroom decorations when reporting a burglary.
